src/main/kotlin/com/jetbrains/snakecharm/lang/psi/impl/SmkUseImpl.kt

```diff
@@ -29,4 +33,71 @@ class SmkUseImpl : SmkRuleLikeImpl<SmkUseStub, SmkUse, SmkRuleOrCheckpointArgsSe
             else -> super.acceptPyVisitor(pyVisitor)
         }
     }
+
+    override fun getNameNode(): ASTNode? {
+        val identifier = super.getNameNode()
+        if (identifier != null) { // Returns new name if we know it
+            return identifier
+        }
+        val namePattern = getNameIdentifierPattern()
+        if (namePattern != null) { // Returns name patter if it exits
+            return namePattern.node
+        }
+        val originalNames = getImportedRuleNames()
+        // Returns original names, we don't want to save one name because
+        // index with this name probably already exists
+        // so we save whole rules names if it is not just '*' wildcard
+        if (originalNames != null && originalNames.text != "*") {
+            return originalNames.node
+        }
+        return null
+    }
+
+    override fun getProducedRulesNames(visitedFiles: MutableSet<PsiFile>): List<Pair<String, PsiElement>> {
+        val identifier = super.getNameNode()
+        if (identifier != null) {
+            return listOf(identifier.text to identifier.psi)
+        }
+        val importedNames = getImportedRuleNames()
+        val newName = getNameIdentifierPattern()
+        if (importedNames == null) {
+            return emptyList()
+        }
+        val originalNames = mutableListOf<Pair<String, PsiElement>>()
+        var child = importedNames.firstChild
+        while (child != null) {
+            if (child.elementType == SmkElementTypes.REFERENCE_EXPRESSION) {
+                originalNames.add(child.text to child)
+            }
+            child = child.nextSibling
+        }
+        if (originalNames.isEmpty()) {
+            val moduleRef = (getModuleName() as? SmkReferenceExpression) ?: return emptyList()
+            val module = moduleRef.reference.resolve()
+            if (module == null || module !is SmkModule) {
+                return emptyList()
+            }
+            val file = module.getPsiFile()
+            if (file != null && file is SmkFile) {
+                originalNames.addAll(file.advancedCollectRules(visitedFiles).map { it.first to it.second })
+            } else {
+                return emptyList()
+            }
+        }
+        return if (newName != null) {
+            originalNames.map { newName.text.replace("*", it.first) to it.second }
+        } else {
+            originalNames.map { it.first.replace("*", it.first) to it.second }
+        }
+    }
+
+    override fun getModuleName() =
+        (findChildByType(SmkTokenTypes.SMK_FROM_KEYWORD) as? PsiElement)?.nextSibling?.nextSibling
+
+    override fun getImportedRuleNames() = findChildByType(SmkElementTypes.USE_IMPORTED_RULES_NAMES) as? PsiElement
+
+    override fun getNameIdentifierPattern() = findChildByType(SmkElementTypes.USE_NAME_IDENTIFIER) as? PsiElement
+
+    override fun containsRuleReference(reference: PsiReference) =
```

Comment:
This method is too specific and is used in a single non-standard case. I think it is ok to inline it and do not provide PSI API for it.

I think `getImportedRuleNames()?.children?` is OK to use in extension and it isn't too low-level as was `node.findChildByType(SmkTokenTypes.SMK_FROM_KEYWORD)?.psi?.nextSibling?.nextSibling` before

Reply:
Done